tests/components/http/test_forwarded.py

```diff
@@ -0,0 +1,487 @@
+
+    mock_api_client = await aiohttp_client(app)
+
+    resp = await mock_api_client.get("/", headers={X_FORWARDED_FOR: x_forwarded_for})
+
+    assert resp.status == 400
+    assert "Invalid IP address in X-Forwarded-For" in caplog.text
+
+
+async def test_x_forwarded_for_with_multiple_headers(aiohttp_client, caplog):
+    """Test that we get a HTTP 400 bad request with multiple headers."""
+    app = web.Application()
+    app.router.add_get("/", mock_handler)
+    async_setup_forwarded(app, [ip_network("127.0.0.1")])
+
+    mock_api_client = await aiohttp_client(app)
+
+    resp = await mock_api_client.get(
+        "/",
+        headers=[
+            (X_FORWARDED_FOR, "222.222.222.222"),
+            (X_FORWARDED_FOR, "123.123.123.123"),
+        ],
+    )
+
+    assert resp.status == 400
+    assert "Too many headers for X-Forwarded-For" in caplog.text
+
+
+async def test_x_forwarded_proto_without_trusted_proxy(aiohttp_client):
+    """Test that proto header is ignored when untrusted."""
+
+    async def handler(request):
+        url = mock_api_client.make_url("/")
+        assert request.host == f"{url.host}:{url.port}"
+        assert request.scheme == "http"
+        assert not request.secure
+        assert request.remote == "127.0.0.1"
+
+        return web.Response()
+
+    app = web.Application()
+    app.router.add_get("/", handler)
+
+    async_setup_forwarded(app, [])
+
+    mock_api_client = await aiohttp_client(app)
+    resp = await mock_api_client.get(
+        "/", headers={X_FORWARDED_FOR: "255.255.255.255", X_FORWARDED_PROTO: "https"}
+    )
+
+    assert resp.status == 200
+
+
+@pytest.mark.parametrize(
+    "x_forwarded_for,remote,x_forwarded_proto,secure",
+    [
+        ("10.10.10.10, 127.0.0.1, 127.0.0.2", "10.10.10.10", "https, http, http", True),
+        ("10.10.10.10, 127.0.0.1, 127.0.0.2", "10.10.10.10", "https,http,http", True),
+        ("10.10.10.10, 127.0.0.1, 127.0.0.2", "10.10.10.10", "http", False),
+        (
+            "10.10.10.10, 127.0.0.1, 127.0.0.2",
+            "10.10.10.10",
+            "http, https, https",
+            False,
+        ),
+        ("10.10.10.10, 127.0.0.1, 127.0.0.2", "10.10.10.10", "https", True),
+        (
+            "255.255.255.255, 10.10.10.10, 127.0.0.1",
+            "10.10.10.10",
+            "http, https, http",
+            True,
+        ),
+        (
+            "255.255.255.255, 10.10.10.10, 127.0.0.1",
+            "10.10.10.10",
+            "https, http, https",
+            False,
+        ),
+        ("255.255.255.255, 10.10.10.10, 127.0.0.1", "10.10.10.10", "https", True),
+    ],
+)
+async def test_x_forwarded_proto_with_trusted_proxy(
+    x_forwarded_for, remote, x_forwarded_proto, secure, aiohttp_client
+):
+    """Test that we get the proto header if proxy is trusted."""
+
+    async def handler(request):
+        assert request.remote == remote
+        assert request.scheme == ("https" if secure else "http")
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5985897)